### ISSUE TITLE
Dask+CuPy crash

### ISSUE THREAD

**Issue by pentschev** [rapidsai/ucx-py#284]
This is a bug that is known by some of us for some weeks now, but I think it was never reported here. I'm opening this with a sample to reproduce the issue:

```python
import time

import dask.array as da
import cupy

from distributed import Client
from dask_cuda import DGX
from dask_cuda.initialize import initialize
from tornado.ioloop import IOLoop

enable_tcp_over_ucx=True
enable_infiniband=False
enable_nvlink=False

async def run():
    initialize(create_cuda_context=True,
               enable_tcp_over_ucx=enable_tcp_over_ucx,
               enable_infiniband=enable_infiniband,
               enable_nvlink=enable_nvlink)

    async with DGX(interface="enp1s0f0",
                   protocol="ucx",
                   enable_tcp_over_ucx=enable_tcp_over_ucx,
                   enable_infiniband=enable_infiniband,
                   enable_nvlink=enable_nvlink,
                   asynchronous=True) as dgx:
        async with Client(dgx, asynchronous=True) as client:
            rs = da.random.RandomState(RandomState=cupy.random.RandomState)
            a = rs.normal(10, 1, (int(4e3), int(4e3)), chunks=(int(1e3), int(1e3)))
            x = a + a.T

            for i in range(100):
                print("Running iteration:", i)
                start = time.time()
                res = await client.compute(x)
                print("Time for iteration", i, ":", time.time() - start)


if __name__ == "__main__":
    loop = IOLoop.current()

    loop.run_sync(run)
```

The example above using CuPy crashes after a few iterations (even zero at times), and seems to be more frequent the slower it runs (i.e., NVLink may run for many more iterations than pure TCP over UCX). The crash ultimately generates the following output:

```python-traceback
[1571346940.363367] [dgx13:33673]          cudamem.c:114  UCX  WARN cuMemGetAddressRange(devPtr=0x7f586e000000) failed
/datasets/pentschev/miniconda3/envs/ucx/bin/python: symbol lookup error: /datasets/pentschev/miniconda3/envs/ucx/lib/ucx/libucm_cuda.so.0: undefined symbol: ucs_fatal_error_format
[1571346940.378122] [dgx13:33550:0]           sock.c:230  UCX  ERROR recv(fd=125) failed: Connection reset by peer
[1571346940.677878] [dgx13:33550:0]           sock.c:230  UCX  ERROR send(fd=-1) failed: Bad file descriptor
distributed.utils - ERROR - Comm Error "[Send #080] ep: 0x7f59a24b50d8, tag: 0x139d77abc337e00a, nbytes: 8": Input/output error
Traceback (most recent call last):
  File "/datasets/pentschev/ucx-src2/distributed/distributed/utils.py", line 662, in log_errors
    yield
  File "/datasets/pentschev/ucx-src2/distributed/distributed/comm/ucx.py", line 124, in write
    await self.ep.send(np.array([len(frames)], dtype=np.uint64))
  File "/datasets/pentschev/miniconda3/envs/ucx/lib/python3.7/site-packages/ucp/public_api.py", line 241, in send
    await self._ep.send(buffer, nbytes=nbytes)
  File "ucp/_libs/core.pyx", line 557, in send
ucp.exceptions.UCXError: Comm Error "[Send #080] ep: 0x7f59a24b50d8, tag: 0x139d77abc337e00a, nbytes: 8": Input/output error
```

**Comment by jakirkham** [comment]
I seem to be getting some different issues. Are there some pieces I'm missing here?

In particular I see the following exception, which gets raised repeatedly for a while.

```python
distributed.protocol.core - CRITICAL - Failed to deserialize
Traceback (most recent call last):
  File "/home/nfs/jkirkham/miniconda/envs/ucx/lib/python3.7/site-packages/distributed/protocol/core.py", line 124, in loads
    value = _deserialize(head, fs, deserializers=deserializers)
  File "/home/nfs/jkirkham/miniconda/envs/ucx/lib/python3.7/site-packages/distributed/protocol/serialize.py", line 268, in deserialize
    return loads(header, frames)
  File "/home/nfs/jkirkham/miniconda/envs/ucx/lib/python3.7/site-packages/distributed/protocol/cuda.py", line 28, in cuda_loads
    return loads(header, frames)
  File "/home/nfs/jkirkham/miniconda/envs/ucx/lib/python3.7/site-packages/distributed/protocol/cupy.py", line 22, in deserialize_cupy_array
    header["shape"], dtype=header["typestr"], memptr=cupy.asarray(frame).data
  File "/home/nfs/jkirkham/miniconda/envs/ucx/lib/python3.7/site-packages/cupy/creation/from_data.py", line 66, in asarray
    return core.array(a, dtype, False, order)
  File "cupy/core/core.pyx", line 1712, in cupy.core.core.array
  File "cupy/core/core.pyx", line 1738, in cupy.core.core.array
  File "cupy/core/core.pyx", line 2518, in cupy.core.core._convert_object_with_cuda_array_interface
TypeError: zip argument #2 must support iteration
```

Eventually I start to see this exception.

```python
distributed.worker - INFO - Dependent not found: ('normal-9df402b6d4f0944fdbc386828f120e39', 3, 2) 0 .  Asking scheduler
distributed.core - ERROR - {'op': 'get_data', 'keys': ("('normal-9df402b6d4f0944fdbc386828f120e39', 3, 2)",), 'who': 'ucx://10.33.227.165:34069', 'max_connections': None, 'reply': True}
Traceback (most recent call last):
  File "/home/nfs/jkirkham/miniconda/envs/ucx/lib/python3.7/site-packages/distributed/core.py", line 403, in handle_comm
    result = await result
  File "/home/nfs/jkirkham/miniconda/envs/ucx/lib/python3.7/site-packages/distributed/worker.py", line 1185, in get_data
    assert response == "OK", response
AssertionError: {'op': 'get_data', 'keys': ("('normal-9df402b6d4f0944fdbc386828f120e39', 3, 2)",), 'who': 'ucx://10.33.227.165:34069', 'max_connections': None, 'reply': True}
```

**Comment by pentschev** [comment]
Ah right. You need to be on latest CuPy and Numba master. This is due to the `__cuda_array_interface__` changes wrt strides. Could you try again after updating?

**Comment by jakirkham** [comment]
Where is Numba getting involved here?

**Comment by quasiben** [comment]
Numba is not involved with CuPy only cuDF

**Comment by quasiben** [comment]
Sorry, misread peter's comment in this thread.

**Comment by jakirkham** [comment]
FWIW I ran something similar with cuDF and got a segmentation fault. Filed it as issue ( https://github.com/rapidsai/ucx-py/issues/290 ) in case it is of interest. Guessing they share the same underlying cause.

**Comment by pentschev** [comment]
@quasiben is correct, I meant you should likely update Numba when testing cuDF, but I haven't tried that, so _maybe_ it isn't needed.

**Comment by jakirkham** [comment]
Sure. I understand that. Just wanted to make sure I wasn't missing something about the CuPy use case. FWIW ucx-py isn't using Numba. So unless dask-cuda is using Numba, I don't think it is needed for your test either.

**Comment by pentschev** [comment]
Neither ucx-py nor dask-cuda _require_ Numba. What requires Numba is cuDF, so that may not work properly without the latest changes to the `__cuda_array_interface__` protocol, but I can't confirm or deny that as of now. So if possible, I would recommend being on Numba 0.46.0 or higher.

**Comment by jakirkham** [comment]
Yep I'm already using the latest release of Numba.

**Comment by quasiben** [comment]
This issue has been resolved upstream in UCX core and will be included in the 1.7rc2 release.  First resolved by https://github.com/openucx/ucx/pull/4431 and later by https://github.com/openucx/ucx/pull/4476
